File: checksum.js
/**
 * Calculating the Checksum for FIX PROTOCOL messages using NodeJS
 *
 * Example messages
 * 8=FIX.4.2|9=49|35=5|34=1|49=ARCA|52=20150916-04:14:05.306|56=TW|
 * Reference : https://stackoverflow.com/questions/32708068/how-to-calculate-checksum-in-fix-manually
 */

var message = process.argv[2];

checksum = 0;
for (let index in message) {
  if (message[index] == '|') {
    checksum += 1;
  } else {
    checksum += message[index].charCodeAt(0);
  }
}

checksum = checksum % 256;
checksum = ('00' + checksum).slice(-3);
message = message + '10=' + checksum + '|';
console.table({
  Input: process.argv[2],
  Output: message,
});

File: checksum.go
/*
Add one message per line as example on messages.txt.
go run checksum.go
Reference : https://stackoverflow.com/questions/32708068/how-to-calculate-checksum-in-fix-manually
*/

package main

import (
	"errors"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
)

func main() {
	readFile()
}

// Read the messages.txt file and send to Checksum calculation function
func readFile() {
	mydir, err := os.Getwd()
	check(err)
	p := filepath.Join(mydir, "messages.txt")
	infos, err := os.ReadFile(p)
	check(err)
	if len(infos) != 0 {
		messages := strings.Split(string(infos), "\r\n")
		calcCheckSum(messages)
	} else {
		err := errors.New("No line in file. Please follow the steps on docs")
		check(err)
	}

}

// Get the slice of strings from the messages.txt file and print on console
func calcCheckSum(args []string) {
	for _, message := range args {
		var soma = 0
		for _, c := range message {
			if string(c) == "|" {
				soma += 1
				continue
			} else {
				soma += int(c)
			}
		}
		fmt.Printf("%s10=%d\n", message, soma%256)
	}
}

// Send an error to this function.
func check(e error) {
	if e != nil {
		log.Fatal(e)
	}
}
